### DISCUSSION TITLE
How to write the correct type for  `useLoaderData()`? (Typescript)

### DISCUSSION THREAD

**Discussion by blatoo** [remix-run/react-router#9792] [Q&A] [answered]
Hi all,

I am follow the official tutorial, but with typescript. I have problem with `useLoaderData()` from here: https://reactrouter.com/en/main/start/tutorial#loading-data

```
  const { contacts } = useLoaderData();
```

The error is at `contacts`, it says: 
**Property 'contacts' does not exist on type 'unknown'.**

So, my question is how to use the useLoaderData() with typescript correctly?

**Comment by msoltan2** [comment, marked as answer]
Just stumbled across the same problem, I temporarily found the next workaround.

```typescript
const data = useLoaderData() as Question[];
```

Where I definitely know that my loader returns an array of type Questions (which is defined by me).

I don't really like this work-around, as it defeats the purpose of typescript and it might introduce bugs.

**Reply by blatoo** [reply]
Hi @Vanilla2 , thanks for the answer. I am also very confused, that the react-router-dom has no official typescript documentation.

**Reply by davinun99** [reply]
It is working for me too @Vanilla2! How can it introduce bugs? Do you think that it has any other drawbacks?

**Reply by jlamoreaux** [reply]
It would be great if it could accept a generic like `const data = useLoaderData<Question>()`

**Reply by MichaelDeBoey** [reply]
@jlamoreaux I implemented that in https://github.com/remix-run/react-router/pull/9670, but we closed it because we first wanted to have https://github.com/remix-run/react-router/discussions/9854 accepted into the roadmap

**Reply by jlamoreaux** [reply]
Amazing 🙌

**Reply by msoltan2** [reply]
> It is working for me too @Vanilla2! How can it introduce bugs? Do you think that it has any other drawbacks?

It can introduce bugs when `useLoaderData()` might not return the same type as the one you are casting to, thus creating a conflict and eventually leading to the component crashing. It is generally advised against using typecasting in typescript unless strictly necessary, as types should be inferred from the return value of the function itself (but in this case this is not possible because the hook itself lacks proper typescript support - generics).

**Reply by jvzaniolo** [reply]
You can do 

```tsx
const whatever = useLoaderData() as ReturnType<typeof loader>
```

**Reply by fredericoo** [reply]
this will get you a promise if you fetch external data, and does not work with `defer` or `json`

**Reply by audunolsen** [reply]
@frederico you can wrap the typecast above with `Awaited` global util-type.

Edit: lmao just saw you posted this in the [thread](https://github.com/remix-run/react-router/discussions/9792#discussioncomment-4809811) doing just that

**Reply by clarencetinator7** [reply]
I have encountered a similar issue but for `useActionData()` and this workaround  seems to work.

**Reply by NoamAnisfeld** [reply]
@Vanilla2 I don't see how generics support would be more typesafe than casting. As long as the type of the data itself is unknown at compile time, all what a generics signature can do is implicit casting.

Assuming the data returned by the loader function is going through some sort of validation at some point, what we actually need is a way to connect to the return type of the loader function.

Perhaps we can go the other way around, and export the expected data type from the same module which we export the component from. Then import it on the router module and force the loader function to comply to that type. I shall try that and post it as an answer if it turns out to be useful.

**Reply by fredericoo** [reply]
@NoamAnisfeld you’re absolutely right in the sense of generics and casting are the same, just disguised as one another. The main advantage of generics here is to filter the types applying serialisation, extracting the value from the promise etc. That’s exactly the sort of utilities I was going for when creating [react-router-typesafe](https://github.com/fredericoo/react-router-typesafe). Just sprinkling the same behaviour of remix onto react router.

**Reply by NoamAnisfeld** [reply]
Thanks @fredericoo, your library is an impressing work. However it's model requires the loader to be part of the component module, which might not always be desired.

**Reply by soullivaneuh** [reply]
> it's model requires the loader to be part of the component module

@NoamAnisfeld Do you mean on the same file? On that case, why should it be required? You can always export your loader from your module then import it on a component to use it with `typeof`. Am I wrong? :thinking:

**Comment by fredericoo** [comment]
I've solved it with a simplified version of Remix's [SerializeFrom](https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/serialize.ts#L76) function:

```ts
import { LoaderFunction } from 'react-router-dom';

export type LoaderData<TLoaderFn extends LoaderFunction> = Awaited<ReturnType<TLoaderFn>> extends Response | infer D
	? D
	: never;
```

usage:

your loader:
```ts
export const loader = (async () => {
	if (FORBIDDEN) return redirect('/signin')
	return { ok: true };
}) satisfies LoaderFunction;
```

Inside your component:
```ts
const data = useLoaderData() as LoaderData<typeof loader>;
```

It will:
✅ Work with data regardless of errors or redirects thrown
✅ Infer your returned data from the loader, without you having to manually type it.

---

EDIT: Released a very thin layer library with everything-typesafe: check out [react-router-typesafe](https://github.com/stargaze-co/react-router-typesafe)

**Reply by fredericoo** [reply]
And… I made a library for this. Don’t bother creating your own patches if yu want a ready-made solution: https://github.com/fredericoo/react-router-typesafe

**Reply by soullivaneuh** [reply]
@fredericoo Your library looks nice, but why not proposing your implementation to the react-router package directly? :thinking:

**Reply by soullivaneuh** [reply]
I got my answer here: https://github.com/remix-run/react-router/discussions/9792#discussioncomment-8439380

May you please share the links of the PR you submitted? Thanks!

**Comment by EdwinChua** [comment]
> Hi all,
> 
> I am follow the official tutorial, but with typescript. I have problem with `useLoaderData()` from here: https://reactrouter.com/en/main/start/tutorial#loading-data
> 
> ```
>   const { contacts } = useLoaderData();
> ```
> 
> The error is at `contacts`, it says: **Property 'contacts' does not exist on type 'unknown'.**
> 
> So, my question is how to use the useLoaderData() with typescript correctly?

### Specific to this tutorial (and OP's question), this is what worked for me:

````ts

    const { contacts } = useLoaderData() as { contacts: [] } || { contacts: [] };

````

The code uses the destructuring syntax on an object that looks like this:

````
{ 
    contacts: [] 
}
````

So providing the exact type, and performing a "short-circuit evaluation" to handle null values will help resolve this.

Further explanations here:

- https://stackoverflow.com/questions/70976866/cannot-destructure-property-as-undefined

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR#short-circuit_evaluation

**Reply by petenilson** [reply]
For those giving a thumbs down please can you explain what you don't like about this solution. I found this discussion also looking for a solution to this problem when completing the tutorial in typescript and I don't have enough experience to know why it's a bad solution.

**Reply by MattyBalaam** [reply]
Can you not use the recommended `useLoaderData<typeof loader>()`

**Reply by fredericoo** [reply]
I downvoted it because the runtime `or` assertion is not good in this instance. `useLoaderData` already falls back to an empty object to not break destructuring or accessing properties. Defaulting to an empty list will make handling loading states harder for no reason, and looks very unreadable — at first I read your code like `as { contacts: [] } | { contacts: [] }` and thought it was a typo, but you simply were casting _and_ setting a fallback via `||`.

@MattyBalaam As of today you cannot do that. Remix has support for generics in loaders, but react-router doesn’t. I even tried to submit a PR to it but the types are very deep and intertwined, and it ended up being very challenging to do without breaking anything. 

For my personal usage, I wrote [react-router-typesafe](https://github.com/fredericoo/react-router-typesafe) but I think it also fixes most ppls issues. It has generics where you’d expect, defer and Await work as you’d expect, and I also made links have editor auto-completion.

**Comment by ryankshaw** [comment]
I remember in early Remix, you had to put those `as` everywhere you used `useLoaderData`, but then the remix team did some really cool fancy typescript work so that when you are in remix-land you can do 

```ts
useLoaderData<typeof loader>()
```

But it looks like the version that they back-ported to `react-router` can't do that. That will give you: `Expected 0 type arguments, but got 1`

to get around it, you can do:
```ts
useLoaderData() as Awaited<ReturnType<typeof loader>>
```

But every time I put an `as` in my code I feel like I'm introducing future lies. @ryanflorence do you know if you plan on adding that typescript fancyness to the react-router version of `useLoaderData`? (as in, make `useLoaderData<typeof loader>()` work like it does in remix?)

**Reply by hoetmaaiers** [reply]
Any update on wether the typescript implementation for `useLoaderData<typeof loader>()` reflows into react-router or not?
My main case for using react-router directly is being able to create client-side-only applications...

Maybe @ryanflorence can chime in the right people to answer this question?

**Reply by fredericoo** [reply]
I know they’re not big fans of this sort of inference, even more so in remix as there’s serialisation needs to cross the backend-frontend boundary. I think Remix v3 will tackle (and fix) most of this, but meanwhile I’ve made a library with minimal patching to ensure react-router works similarly to Remix: https://github.com/fredericoo/react-router-typesafe

**Reply by eric-horodyski** [reply]
@fredericoo - Is there a rationale behind creating a "fork" of react-router instead of contributing to a types package? Something like `@types/react-router`?

**Reply by fredericoo** [reply]
> @fredericoo - Is there a rationale behind creating a "fork" of react-router instead of contributing to a types package? Something like `@types/react-router`?

react router _is_ typed; Loader and action functions, json and defer however all return unknown or non inferred types. My library is a thin layer on top of these bringing generics passed as arguments and a bit more inference in the case of json and deferred. The aim is for it to behave closer to the way remix works. 

The Remix team has actively said this is by design and they're still not happy with how remix implements it today. As it's a blocker for me in several projects, I've decided on publishing the library. 

It's not typings on top of an untyped library, they're re-exports under different signatures to make developers' lives easier until we have a permanent, official solution. Try it out! I'm about to publish a patch to the Await component as well with the desired inference ◡̈

**Reply by eric-horodyski** [reply]
I see! Apologies for the misunderstanding, bouncing around different languages/frameworks/contexts doesn't always lend itself to asking the most accurate questions.

**Reply by LucasKuhn** [reply]
> [Carl Sagan](https://en.wikipedia.org/wiki/Carl_Sagan), in his work [The Demon-Haunted World: Science as a Candle in the Dark](https://en.wikipedia.org/wiki/The_Demon-Haunted_World:_Science_as_a_Candle_in_the_Dark) said: "There are naïve questions, tedious questions, ill-phrased questions, questions put after inadequate self-criticism. But every question is a cry to understand the world. There is no such thing as a dumb question".

Thank you both for the question and the answer, it made the current situation with react-router clear and the reasoning for the separate library 🙏

**Comment by rafaelbiten** [comment]
Until we have the generics for these fns, I decided to go with a helper fn that wraps the default `useLoaderData`:

```ts
function useDataFromLoader<LoaderFn extends LoaderFunction>(loaderFn: LoaderFn) {
  return useLoaderData() as Awaited<ReturnType<typeof loaderFn>>
}
```

Usage:
```ts
async function loader() {
  const data = await Promise.resolve({ foo: 'bar' })
  return data
}  satisfies LoaderFunction

function Component() {
  const { foo } = useDataFromLoader(loader)
  return <p>foo === 'bar': {foo === 'bar'}</p>
}
```

For the `useRouteLoaderData`, I used a similar approach:

```ts
export function useDataFromRouteLoader<LoaderFn extends LoaderFunction>(route: { id: string; loader: LoaderFn }) {
  return useRouteLoaderData(route.id) as Awaited<ReturnType<typeof route.loader>>
}
```

Usage:
```ts
const data = useDataFromRouteLoader(routeWithLoaderAndId)
```

I personally find this to be a simple solution to the problem, but I don't know if it's good enough to be part of the public API of the library, e.g.: have the `useLoaderData` take a loader fn just so it can figure out its return type...

**Reply by fredericoo** [reply]
bro this will ruin code splitting! your loader now is going to ship in the same chunk as your component code, which mostly isn't what you want. 
Avoid importing the loader itself in your component if you want to code split your app. That's where `typeof` shines.

**Comment by mohsentaleb** [comment]
`Awaited<ReturnType<typeof loaderFn>>` can't always be the answer since `loader` isn't always with the signature `(args: LoaderFunctionArgs) => unknown`.  It can sometimes be `(curriedFunction) => (args: LoaderFunctionArgs) => unknown`

Let me give you an example:
A common use case for a `loader` function is to fetch data from API which needs user's `accessToken` being available in the scope of that function. like so:
```ts
axios.get('/some/api/endpoint', {
    headers: {
        AUTHORIZATION: `Bearer ${authData.token}`
    }
})
```

After some digging and with the help of [this stackoverflow question](https://stackoverflow.com/questions/74296504/invoke-hook-from-react-router-data-router-action/74296637#74296637) I came up with the following workaround:
```ts
// App.tsx
export default function App() {
    const { authData } = useAuth(); // A simple hook which helps us get access to some app Context.

    const apiClient = axios.create({
        baseURL: 'https://domain.com',
        headers: {
            AUTHORIZATION: `Bearer ${authData.token}`
        },
    });

    const router = createBrowserRouter(
        createRoutesFromElements(
            <Route path="content" element={<Component />} loader={contentLoader(apiClient)} />
        )
    );

    return <RouterProvider router={router} />;
}
```
And then in my Component:
```ts
// Content.tsx

// loader as a curried function
export const loader =
    (apiClient: AxiosInstance) =>
    async ({ params }: LoaderFunctionArgs) => {
        const contentInfo = await apiClient.get<Content>(
            `content/${params.id}`
        );

        return contentInfo;
    };
```

Now if we use `Awaited<ReturnType<typeof loader>>` we get the second curried function as the type which is `async ({ params }: LoaderFunctionArgs) => { ...}` and not the `contentIndo`.

**Reply by rafaelbiten** [reply]
I haven't tested this myself, but by looking at your code and that SO, the approach of wrapping `useLoaderData` into another function should still work, e.g.:

```ts
const apiClient = axios.create(...)
const loaderWithApi = loader(apiClient)

...

function Component() {
  // using the useDataFromLoader wrapper I suggested on my previous message
  // and typescript should be able to infer what contentInfo info is
  const contentInfo = useDataFromLoader(loaderWithApi)
  return <pre>{JSON.stringify(contentInfo, null, 2) }</pre>
}
```

You may have to refactor your code a bit for it to work. I'm still learning about the new router and looking for the pattern I like. So far, I have the basics set like so:

```ts
export const myRoute = {
  element: <Element />,
  errorElement: <ErrorElement />,
  loader: async function ({ params }) {
    if (!params.id) return { data: null }
    const data = await api.getSomethingWithId(params.id)
    return { data }
  } satisfies LoaderFunction,
}

// in the same file, no need to export
function Element() {
  const { id } = useParams()
  // using the useDataFromLoader wrapper I suggested on my previous message
  // Typescript should be able to infer that `data` is either `Something | null` (in this example)
  const { data } = useDataFromLoader(contactShowRoute.loader)
  ...
```

Then, when declaring the router:
```ts
const router = createBrowserRouter([
  {
    id: 'myPath', // id may be helpful in some cases: https://reactrouter.com/en/main/hooks/use-route-loader-data
    path: '/myPath',
    ...myRoute, // contains elements, loader, action...
    children: [ ... ] // keeping the children explicit
  }
])
```

It may be just a question of time before I find some drawbacks with this approach, but that's what I'm using so far and it works well with that little `useLoaderData` wrapper I wrote...

**Comment by seanmcquaid** [comment]
Does anyone have a solution for typing this when working with defer?? My current solution is to manually create an interface for the desired types when working with defer but would love it if we had a supported TS solution for using this hook from the maintainers if possible!

**Reply by fredericoo** [reply]
Hey! I. put together a library for all things react-router + typescript: https://github.com/fredericoo/react-router-typesafe

**Comment by ivosabev** [comment]
The lack of automated framework type safety between loaders/action and the route component is a big turn off at the moment. Could the core team address this asap?

**Reply by fredericoo** [reply]
I put this library together but cant wait for it to be deprecated: https://github.com/fredericoo/react-router-typesafe

**Reply by eric-horodyski** [reply]
Totally agree. I'm advocating the move to React Router 6 internally, but struggling because our framework is heavily geared towards TypeScript developers.

**Comment by mohsentaleb** [comment]
This would get even worse when you use `defer` in `loader`.
For a simple `loader` you could use `const data = useLoaderData() as ReturnType<typeof loader>>` but as soon as you use `defer` for your loader's return types are completely lost.

**Reply by fredericoo** [reply]
I’ve patched defer with my own type definition, maybe I’ll release a library for these react router typed methods, which work well for me!

**Reply by QuentinGibson** [reply]
@fredericoo how did you patch defer? I am struggling with it!

**Reply by fredericoo** [reply]
I ended up making a library, feel free to copy the source code or install it (I’ll maintain it as I’m using it in several projects)!

https://github.com/fredericoo/react-router-typesafe

**Comment by brandon-burciaga** [comment]
Bumping this as generic support seems pretty necessary here.

**Reply by fredericoo** [reply]
I tried writing a PR to improve generic support but it's VERY HARD. There are so many intertwined types I felt like I was breaking the whole thing.

**Reply by fredericoo** [reply]
Ended up making a library for these repeated uses: https://github.com/fredericoo/react-router-typesafe

**Reply by LucasKuhn** [reply]
Would love to see the oficial docs with a Typescript version of the tutorial

**Comment by ntkoopman** [comment]
Depending on what you think is acceptable, you could skip `useLoaderData`, and return a component from the loader itself:

```tsx
const Users = ({users}) => {
  return (
    <div>
      {users.length} users
    </div>
  );
}

const loader = async () => {
  let users = await db.select.from(users);
  return <Users users={users} />
}

// You can reuse this component on every route
const Passthrough = () => useLoaderData() as JSX.Element;

const router = createBrowserRouter([
  {
    path: "/users",
    element: <Passthrough />,
    loader: loader,
  },
]);
```

**Comment by tar-aldev** [comment]
Imho it looks a bit weird that we don't have it in react-router-dom lib, given that remix has it.
https://remix.run/docs/en/main/start/tutorial#type-inference

The tutorials are quite similar, the unerlying routing is (probably?)  quite simlar. 
So not sure why we can't have the same consistent approach in both remix and react-router.

**Reply by kiliman** [reply]
Most likely because React Router doesn't have a network request between the loader and component. With Remix, the loader data is converted to a JSON payload, whereas with React Router, you return exactly what you get from `useLoaderData`.

**Comment by madfcat** [comment]
Is it not solved yet?
How can I get the right type for   `const data = useLoaderData<typeof loader>();` after having my graphql called in `loader()`?

**Reply by fredericoo** [reply]
with react router 7 you can just use the type generated routes and access `loaderData` straight from your route component props, but if you want you can also pass the generic as you just did. 

what’s important is that you don’t annotate the return type of the loader.

e.g.:
```ts
import type { Route } from './+types/_index';

export async function loader({request}: Route.LoaderArgs) {
  return { orders: awaitgetOrders() }
}

export default function Route({loaderData}: Route.ComponentProps) {
  return <div>{loaderData.orders.length} orders</div>
}
```

same works for clientLoaders and actions

**Reply by madfcat** [reply]
I guess it is not available in Remix 2 used in Shopify's app scaffold template. That's what I am using (`2.16.1`)

**Reply by fredericoo** [reply]
with remix (or react router 7) you can just pass the generic like so:
```ts
export default function Route() {
    const loaderData = useLoaderData<typeof loader>;
}
```

it’s important that you let the return type of the loader be inferred or deliberately annotate it

**Reply by madfcat** [reply]
> ```ts
> export default function Route() {
>     const loaderData = useLoaderData<typeof loader>;
> }
> ```

`loaderData` becomes of type `any` and loses any intellisense type suggestion in vscode.

**Reply by madfcat** [reply]
Update: `LoaderArgs` did its job.

I still can not type `extensions`, `headers` that I am getting from the graphql response from Shopify.
It comes as 
```
{ data: ...,
extensions: ...,
headers: ...
}
```

I am still not sure how to get types out of that. For example for:
```
00:42:16 │         remix │   extensions: {
00:42:16 │         remix │     cost: {
00:42:16 │         remix │       requestedQueryCost: 74,
00:42:16 │         remix │       actualQueryCost: 23,
00:42:16 │         remix │       throttleStatus: [Object]
00:42:16 │         remix │     }
00:42:16 │         remix │   },
```

Another problem is of course `JsonifyObject<>` which wraps my responseJson type 🤦‍♀️
